### PR TITLE
-resolve instruction extended with beforelaunch

### PR DESCRIPTION
**Pull request by pkriens**
This PR extends the `-resolve` instruction from `manual` (default) and `auto` with `beforelaunch`. The `beforelaunch` option will  calculate the `-runbundles` on demand. It has a cache based on the project properties.

### CODE REVIEW

File: biz.aQute.bndlib/src/aQute/bnd/service/result/FunctionWithException.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
+package aQute.bnd.service.result;
```
Comment (rotty3000) [inline biz.aQute.bndlib/src/aQute/bnd/service/result/FunctionWithException.java]: We already have https://github.com/bndtools/bnd/blob/master/aQute.libg/src/aQute/lib/exceptions/FunctionWithException.java

File: biz.aQute.resolve/src/biz/aQute/resolve/RunResolution.java [hunk 2]
```diff
@@ -354,4 +390,122 @@ public String report(boolean optionals) {
...
+
+		try {
+			File f = getCacheFile(project);
+
+			CacheDTO dto = new CacheDTO();
+			dto.checksum = project.getChecksum();
+			dto.runbundles = getRunBundles();
+
+			Path tmp = Files.createTempFile(f.getParentFile()
+				.toPath(), project.getName(), ".json");
```
Comment (rotty3000) [inline biz.aQute.resolve/src/biz/aQute/resolve/RunResolution.java]: This gave an error in windows (see azure pipeline failure):

```
biz.aQute.resolve.RunResolutionTest > testLaunchWithBeforeLaunchResolve FAILED
    java.nio.file.InvalidPathException: Illegal char <:> at index 71: C:\Users\VssAdministrator\.bnd\default-ws\cnf\cache\4.4.0\resolutions\D:\a\1\s\biz.aQute.resolve\testdata\enroute
        at sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
        at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
        at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
        at sun.nio.fs.WindowsPath.parse(WindowsPath.java:94)
        at sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:255)
        at java.io.File.toPath(File.java:2234)
        at biz.aQute.resolve.RunResolution.cache(RunResolution.java:416)
        at biz.aQute.resolve.RunResolution.lambda$getRunBundles$4(RunResolution.java:469)
        at aQute.bnd.service.result.Ok.then(Ok.java:113)
        at biz.aQute.resolve.RunResolution.getRunBundles(RunResolution.java:467)
        at biz.aQute.resolve.Bndrun.getRunbundles(Bndrun.java:154)
        at aQute.bnd.build.ProjectLauncher.updateFromProject(ProjectLauncher.java:123)
        at aQute.launcher.plugin.ProjectLauncherImpl.updateFromProject(ProjectLauncherImpl.java:92)
        at aQute.bnd.build.Project.getProjectLauncher(Project.java:2818)
        at biz.aQute.resolve.RunResolutionTest.testLaunchWithBeforeLaunchResolve(RunResolutionTest.java:113)
```

Comment (rotty3000) [reply biz.aQute.resolve/src/biz/aQute/resolve/RunResolution.java]: https://dev.azure.com/bndtools/bnd/_build/results?buildId=1843&view=logs&j=5d1700ad-7d0a-54b3-d968-60bb514cf4d6&t=f44c40b5-a7fb-58cc-8f9d-bc280c1c5d20&l=395

File: biz.aQute.bndlib/src/aQute/bnd/help/instructions/ResolutionInstructions.java [hunk 3]
```diff
@@ -51,4 +55,25 @@
 	@SyntaxAnnotation(lead = "Specify the runorder and startlevel behavior of the resolved bundles", example = "'-runstartlevel order=leastdependenciesfirst, begin=1, step=1")
 	RunStartLevel runstartlevel(RunStartLevel deflt);
 
+	enum ResolveMode {
+		/**
+		 * Resolve is done manually. This is the default
+		 */
+		@SyntaxAnnotation(lead = "Resolve is manyally with the Resolve button")
```
Comment (rotty3000) [inline biz.aQute.bndlib/src/aQute/bnd/help/instructions/ResolutionInstructions.java]: typos?

File: biz.aQute.bndlib/src/aQute/bnd/osgi/Constants.java [hunk 4]
```diff
@@ -204,6 +204,12 @@
 	String				STRICT										= "-strict";
 	String				SUB											= "-sub";
 	String				REPRODUCIBLE								= "-reproducible";
+
+	String				RESOLVE										= "-resolve";
+	String				RESOLVE_MANUAL								= "manual";
+	String				RESOLVE_AUTO								= "auto";
+	String				RESOLVE_ONLAUNCH							= "onlaunch";
```
Comment (rotty3000) [inline biz.aQute.bndlib/src/aQute/bnd/osgi/Constants.java]: Are these values supposed to align with `ResolutionInstructions.ResolveMode`? `onlaunch=beforelaunch`?

**Comment by rotty3000** [comment]
I think I get the flow. I do find that I have issues when using the bnd-run-maven-plugin with this. It does not seem to be populating the bundles in a way that the launcher will update them in the running instance although it does do a re-resolve each time a change is detected. Sometimes the resolve cache kicks in even when it should be an update. Maybe we can talk through it on our next call.

**Comment by bjhargrave** [comment]
Did you mean to force push the commit for PR #3583 into this PR? All the previous commits to this PR are gone.

**Comment by pkriens** [comment]
I think I got github confused or the other way around. I saw it was confused, thn created a new PR but somehow it still used the old. I will reopen another PR for the resolve.

File: biz.aQute.bndlib/src/aQute/bnd/service/result/FunctionWithException.java (hunk 1 above)
Comment (pkriens) [reply biz.aQute.bndlib/src/aQute/bnd/service/result/FunctionWithException.java]: The libg is private, it cannot be used in public API

**Comment by bjhargrave** [comment]
> I saw it was confused

I think the issue is that you were using your master branch as the branch for a PR. This always seems to be problematic. Best to always is a specific branch for each PR.

**Pull request merged**